File: go/2.add-two-numbers.go
package leetcode

/*
 * @lc app=leetcode id=2 lang=golang
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func addTwoNumbers(l1 *ListNode, l2 *ListNode) *ListNode {
    head := &ListNode{}
	p := head
	carry := 0
	for l1 != nil || l2 != nil {
		var v1, v2 int
		if l1 != nil {
			v1 = l1.Val
			l1 = l1.Next
		}
		if l2 != nil {
			v2 = l2.Val
			l2 = l2.Next
		}
		sum := v1 + v2 + carry
		carry = sum / 10
		p.Next = &ListNode{Val: sum % 10}
		p = p.Next
	}
	if carry > 0 {
		p.Next = &ListNode{Val: carry}
	}
	return head.Next
}

// @lc code=end

File: 2.add-two-numbers.swift
/*
 * @lc app=leetcode id=2 lang=swift
 *
 * [2] Add Two Numbers
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     public var val: Int
 *     public var next: ListNode?
 *     public init() { self.val = 0; self.next = nil; }
 *     public init(_ val: Int) { self.val = val; self.next = nil; }
 *     public init(_ val: Int, _ next: ListNode?) { self.val = val; self.next = next; }
 * }
 */
class Solution {
    func(_ l1: ListNode?, _ l2: ListNode?) -> ListNode? {
        var l1 = l1
        var l2 = l2
        var head: ListNode? = nil
        var tail: ListNode? = nil
        var carry = 0
        while l1 != nil || l2 != nil {
            let sum = (l1?.val ?? 0) + (l2?.val ?? 0) + carry
            carry = sum / 10
            let node = ListNode(sum % 10)
            if head == nil {
                head = node
            } else {
                tail?.next = node
            }
            tail = node
            l1 = l1?.next
            l2 = l2?.next
        }
        if carry > 0 {
            tail?.next = ListNode(carry)
        }
        return head
    }
}
// @lc code=end

